### ISSUE TITLE
BUG mnlogit does not store design info, predict fails with formula

### ISSUE THREAD

**Issue by eXcuvator** [statsmodels/statsmodels#3958]
Can't predict in mnlogit, test-code follows. Version: `0.8.0`

Independent SE question (author not me): https://stackoverflow.com/questions/44199054/how-to-specify-a-prediction-using-statsmodels-mnlogit-model

```
import statsmodels.formula.api as sm
import pandas as pd
df = pd.DataFrame({"A": [0, 1, 0, 0, 1], "B": [0, 1, 0, 1, 1]})
result = sm.mnlogit(formula="A ~ B", data=df).fit()
result.predict()  # works

# the following lead to 
# AttributeError: 'PandasData' object has no attribute 'design_info'

result.predict(df)
result.predict(df['B'])

```

**Comment by josef-pkt** [comment]
Thanks for the report.

I get the same exception in master running the example as provided.

We refactored the location of design_info at some point to avoid similar problems, but maybe MNLogit has special code that doesn't do the same thing as the generic data and formula handling.

**Comment by josef-pkt** [comment]
To the description title: I don't see any pickling involved here. @eXcuvator  ?

(The original refactoring of the design_info location was to avoid attaching anything directly to a dataframe)

**Comment by josef-pkt** [comment]
current guess: `MultinomialModel(BinaryModel)` overrides  `_handle_data` and does not call `super`. So something might be missing there.
I don't see any special `from_formula` code in discrete_models.

**Comment by josef-pkt** [comment]
I see only a single unit test for mnlogit with formula which is in test_discrete test_mnlogit_factor. 
So my guess would be that this never worked, i.e. formula support where MNLogit deviates from generic code.

**Comment by josef-pkt** [comment]
got it, I think

at the end of `handle_data`, extra keys/kwargs are `pop`ed
base.model excludes design_info and formula, MNLogit does not
```
        for key in kwargs:
            if key in ['design_info', 'formula']:  # leave attached to data
                continue
```

After adding this to MNLogit, the example runs without exception. (but still a covergence warning maxiter exceeded)

**Comment by josef-pkt** [comment]
I'm opening a PR with the fix, but my guess is that calling super would be the better solution, but refactorings need better unit test coverage.

**Comment by josef-pkt** [comment]
temporary workaround for predict
If the entire design matrix for predict, exog_predict, is available, then `transform` can be set to False, which avoids going through patsy and the missing design_info

```
>>> from statsmodels.tools import add_constant
>>> result.predict(add_constant(df['B']), transform=False)
          0             1
0  1.000000  1.401148e-11
1  0.333333  6.666667e-01
2  1.000000  1.401148e-11
3  0.333333  6.666667e-01
4  0.333333  6.666667e-01
```

**Comment by eXcuvator** [comment]
@josef-pkt  Yes, pickling not directly, I took that from related bugs from 2015. Great that there is already a quick work around!

**Comment by josef-pkt** [comment]
The main problem with the workaround is that users need to handle the transforms themselves if transforms are needed. Handling those automatically is the main advantage of patsy (e.g. for categoricals).
Doing this on the user side for exog and for predict is a bit tedious.

fix in PR #3959

**Comment by eXcuvator** [comment]
Oh that's actually right, my "real" case has fixed effects`A ~ C(B)`. I guess I'll be eagerly awaiting the next release in that case.

**Comment by josef-pkt** [comment]
If you have just one set of fixed effects, then using pandas to_dummies and avoiding formulas for now might be the simpler work around. (I have no working experience with pandas categoricals, so just a guess.)
The fix is just two lines, in case you want to edit your statsmodels source, or use master once the fix is merged, most likely later today.

**Comment by josef-pkt** [comment]
I changed the title of the issue

The bug here is independent of pickling. Some background:

note the question referenced above is for mnlogit
https://stackoverflow.com/questions/44199054/how-to-specify-a-prediction-using-statsmodels-mnlogit-model
with workaround (same as here) in my answer at
https://stackoverflow.com/questions/20724919/pandas-dataframe-attributeerror-dataframe-object-has-no-attribute-design-inf
which referred to pickling with formulas in general

The initial changes, after #1263, fixed the design_info problem for pickling (and workaround for missing pickling support in patsy), but we never used MNLogit as a test case.

**Comment by josef-pkt** [comment]
@eXcuvator Thanks again for reporting

I just merged the fix in #3959
